### DISCUSSION TITLE
Is there an easy way to marshal to unicode instead of ansi in generated code?

### DISCUSSION THREAD

**Discussion by kmgallahan** [SharpGenTools/SharpGenTools#204] [Q&A] [answered]
This header function:

https://github.com/kmgallahan/AllSystemsGo.SonyCameraRemoteSDK_SGT/blob/6d89c003ea303c53b31a8df127276c60ced1f3d6/AllSystemsGo.SonyCameraRemoteSDK/CrSDK/CameraRemote_SDK.h#L121

Produces this code:

```
public static unsafe int SetSaveInfo(long deviceHandle, string path, string prefix, int no)
        {
            System.IntPtr path_;
            System.IntPtr prefix_;
            int __result__;
            path_ = System.Runtime.InteropServices.Marshal.StringToHGlobalAnsi(path);
            prefix_ = System.Runtime.InteropServices.Marshal.StringToHGlobalAnsi(prefix);
            __result__ = SetSaveInfo_(deviceHandle, (void *)path_, (void *)prefix_, no);
            System.Runtime.InteropServices.Marshal.FreeHGlobal(path_);
            System.Runtime.InteropServices.Marshal.FreeHGlobal(prefix_);
            return __result__;
        }
```

But I'm getting an error code returned when trying to call the function. The documentation states:

> Camera Remote SDK works in Unicode, the folder path and the prefix must be set as Unicode string. 

So I'm wondering if Marshal.StringToHGlobalAnsi is causing me problems, and if there is a straightforward way to have the generated code use Marshal.StringToHGlobalUni instead?

**Comment by andrew-boyarshin** [comment, marked as answer]
It appears either `_UNICODE` or `UNICODE` preprocessor definitions are not set when parsing `CrTypes.h`, so `CrChar` is `CrAChar` (`char`). SharpGen does the correct thing by assuming it's ANSI, not Unicode. I'm not sure if `UNICODE` not being defined is SharpGen's fault, but you can try adding that preprocessor definition like this:
<https://github.com/SharpGenTools/SharpGenTools/blob/3f1c00ce5883ede01c87ea5c2ea398efde667e61/SharpGenTools.Sdk/Sdk.targets#L85-L87>

**Reply by kmgallahan** [reply]
Passing that argument verbatim didn't change the output, but I dug a bit and found this CastXML arg:

```
-D <macro>=<value>
Define <macro> to <value> (or 1 if <value> omitted)
```

and passing `<CastXmlArg Include="-D=UNICODE" />` updates all generated files to use unicode.

**Reply by jkoritzinsky** [reply]
You can also use a `<define>` element in your mapping to define the UNICODE macro.

**Reply by kmgallahan** [reply]
@jkoritzinsky What is the syntax for that? Can't quite figure it out from [the docs](https://sharpgentools.readthedocs.io/en/latest/custom-mapping.html). Keep in mind I'm not a cpp user and am a novice with SharpGenTools.

**Comment by jkoritzinsky** [comment]
Looks like I mis-remembered. There is a way to do this though with an include prolog.

In your `include` tag for your header, you can add a `pre` element that includes some C++ before the header is included, like so:

```
<include file="Foo.h" attach="true">
   <pre>
        #define UNICODE 1
    </pre>
</include>
```

________________________________

**Reply by kmgallahan** [reply]
Alright, I like that a bit better as it keeps things to the mapping rather than getting into the project file. Thanks.
